### ISSUE TITLE
[Select] [Vue]: Slot props / Default Slot

### ISSUE THREAD

**Issue by jiblett1000** [chakra-ui/ark#826]
My onslaught continues :)

It would be great to see some more (any?) slot props in the select component. Something along the lines of:

```html
  <Select v-model="selectedOption" v-slot="{ isOpen }">
    <SelectLabel>Framework:</SelectLabel>
    <SelectTrigger>
      <button>{{ selectedOption?.label ?? 'Select' }}</button>
    </SelectTrigger>

    <Teleport to="body">
      <SelectPositioner>
        <SelectContent>
          <SelectOption 
            v-for="option in options" 
            :key="option.value" 
            v-slot="{ highlighted, selected, disabled }">
               <span>{{ option.label }}</span>
               <span v-if="selected">
                 <CheckIcon  />
               </span>
          </SelectOption>
        </SelectContent>
      </SelectPositioner>
    </Teleport>
  </Select>
```

While a lot of this can be achieved by using a template ref on the root element and using the api through that, I believe this is much better in terms of DX. Notably this adds the `isOpen` slot prop on the `<Select>` component as well as `highlighted, active, disabled` slot props on the `SelectOption` component.

Another thing of note is being able to pass in custom html through the default slot of `SelectOption` rather than the currently required props of `label` and `value`. This allows the addition of a checkmark (for example) for the selected item or whatever else the dev desires.

I'm trying to migrate from using `headlessui`. The DX of their components is great and provides a great example in general of what I'm talking about: https://headlessui.com/vue/listbox#listbox-option. Unfortunately, they don't seem particularly interested in adding more components.

But, I like the concept of Ark, Zag, etc which is why I'm trying to switch over.

**Comment by TylerAPfledderer** [comment]
Hello again, @jiblett1000!

Firstly, noting that this is an extension of Discussion Post #788, so thank you for your thoroughness here!

Yes, I will address exposing state for the Select via `v-slot`. It's a shame this directive does not have TypeScript support without specifying the type like `v-slot="{ isOpen } : { isOpen: boolean }"`. But it appears that it is being considered in a Vue RFC for the future.

We have type signatures in the package for the api object, but are `ComputedRef` objects which might cause some pain in a use case like this or with a template ref if we use `expose`. I am looking possibly changing this to expose a type that is just a plain object. (This is a part of a more global cleanup for the Vue package)

In short for now, I will at the very least add the v-slot option for `Select` and expose a `SelectContext` type that is not wrapped in a Ref type (if that makes sense??)

Your suggestion above in simply exposing api props like `v-slot="{ isOpen }"` can work here.

---

As to the `SelectOption` and not accepting custom HTML currently: this is actually a bug, as this component was originally intended to accept children instead of rendering the `label`. So this should be a quick fix! 😄 

Let me know how either of these work for you. I am here for the additional input! 🤘🏼

**Comment by jiblett1000** [comment]
@TylerAPfledderer Thanks for digging into it. Bummer Vue doesn't support typing slot props yet, but as long as it's on the horizon, that's good enough for me :) . Good to hear about `SelectOption`. And in regards to making the slot props (selected, highlighted, disabled) available on that component; will that be possible using the same method as `isOpen`? Or not until the fix in Vue?

**Comment by TylerAPfledderer** [comment]
> And in regards to making the slot props (selected, highlighted, disabled) available on that component; will that be possible using the same method as isOpen?

Yes! They basically come from the Zag Select api object. Just dumping that object into the slot for now. 😁

**Comment by TylerAPfledderer** [comment]
@jiblett1000 I think it's worth clarifying the above comment.

This is the [api object](https://github.com/chakra-ui/zag/blob/0a3b173f07941b4f07b78f298caa62395caa7282/packages/machines/select/src/select.connect.ts#L35) that will get exposed through the slot prop.

Maybe what you are looking for comes from the method [getOptionState](https://github.com/chakra-ui/zag/blob/0a3b173f07941b4f07b78f298caa62395caa7282/packages/machines/select/src/select.connect.ts#L22-L29) that is in this object?

**Comment by jiblett1000** [comment]
@TylerAPfledderer yeah getOptionState seems to be it

**Comment by TylerAPfledderer** [comment]
@jiblett1000 PR now merged, give it a try and let me know what you think! 😄

**Comment by jiblett1000** [comment]
@TylerAPfledderer Thank you! Will do. BTW, is there an edge channel?

**Comment by TylerAPfledderer** [comment]
@jiblett1000 do you mean a Discord channel?

If so, there's not currently one available at this time. I would expect one to be made in the future once Ark is released into beta.

**Comment by TylerAPfledderer** [comment]
@jiblett1000 ahh! You can access the latest changes via installing `@ark-ui/vue@0.0.0-rc-20230428162935`

**Comment by jiblett1000** [comment]
@TylerAPfledderer Awesome. Thanks! I checked yesterday but the latest one was from like 12 days ago. I guess it hadn't been reeleased just yet. I'll give it a go :smile:

**Comment by TylerAPfledderer** [comment]
@jiblett1000 yea, full releases might be slow right now 😅

**Comment by jiblett1000** [comment]
@TylerAPfledderer The `isOpen` slot prop seems to be working as I would expect. I can remove my workaround for that now. :+1: 

Perhaps I was misunderstanding in regards to `SelectOption` though. This is the functionality I was thinking of (if possible):

```vue
<SelectOption
    v-for="{ label, value } in options"
    v-slot="{ selected, highlighted, disabled }" // { selected: boolean, highlighted: boolean, disabled: boolean }
    :key="value"
    :label="value"
    :value="value"
    :class="[
        'flex items-center justify-between gap-x-4 px-4 py-3',
        highlighted
            ? 'bg-secondary-container-light text-on-secondary-container-light'
            : selected
            ? 'bg-surface-4-light'
            : 'bg-surface-3-light',
    ]"
>
    <span>{{ label }}</span>
    <CheckIcon
        v-if="selected"
        class="w-6 h-6"
     />
</SelectOption>
```


I also tried this on the `Select` component:

```vue
<Select ref="selectEl" v-model="selected" v-slot="{ isOpen, selectedOption, highlightedOption }">    
```

And then this in `SelectOption`:

```vue
<SelectOption
    v-for="{ label, value } in options"
    :key="value"
    :label="value"
    :value="value"
    :class="[
        'flex items-center justify-between gap-x-4 px-4 py-3',
        highlightedOption?.label === label
            ? 'bg-secondary-container-light text-on-secondary-container-light'
            : selectedOption?.label === label
            ? 'bg-surface-4-light'
            : 'bg-surface-3-light',
    ]"
>
    <span>{{ label }}</span>
    <CheckIcon
        v-if="selectedOption?.label === label"
        class="w-6 h-6"
     />
</SelectOption>
```

While I could see the correct selected option by adding this somewhere in the template `{{ selectedOption }}`; the comparisons I made in `SelectOption` (i.e. `selectedOption?.label === label`)had no effect.

**Comment by jiblett1000** [comment]
I'm seeing now how to use `getOptionState`. But, is there a way to expose those booleans as slot props like in my first example above? I think it would be much better in terms of DX.
